alacritty_terminal/src/ansi.rs

```diff
@@ -81,15 +97,157 @@ fn parse_number(input: &[u8]) -> Option<u8> {
+}
+
 /// The processor wraps a `vte::Parser` to ultimately call methods on a Handler.
+#[derive(Default)]
 pub struct Processor {
     state: ProcessorState,
     parser: vte::Parser,
 }
 
-/// Internal state for VTE processor.
-struct ProcessorState {
-    preceding_char: Option<char>,
+impl Processor {
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Process a new byte from the PTY.
+    #[inline]
+    pub fn advance<H, W>(&mut self, handler: &mut H, byte: u8, writer: &mut W)
+    where
+        H: Handler,
+        W: io::Write,
+    {
+        if self.state.sync_state.timeout.is_none() {
+            let mut performer = Performer::new(&mut self.state, handler, writer);
+            self.parser.advance(&mut performer, byte);
+        } else {
+            self.advance_sync(handler, byte, writer);
+        }
+    }
+
+    /// End a synchronized update.
+    pub fn stop_sync<H, W>(&mut self, handler: &mut H, writer: &mut W)
+    where
+        H: Handler,
+        W: io::Write,
+    {
+        // Process all synchronized bytes.
+        for i in 0..self.state.sync_state.buffer.len() {
+            let byte = self.state.sync_state.buffer[i];
+            let mut performer = Performer::new(&mut self.state, handler, writer);
+            self.parser.advance(&mut performer, byte);
+        }
+
+        // Resetting state after processing makes sure we don't interpret buffered sync escapes.
+        self.state.sync_state.buffer.clear();
+        self.state.sync_state.timeout = None;
+    }
+
+    /// Synchronized update expiration time.
+    #[inline]
+    pub fn sync_timeout(&self) -> Option<&Instant> {
+        self.state.sync_state.timeout.as_ref()
+    }
+
+    /// Number of bytes in the synchronization buffer.
+    #[inline]
+    pub fn sync_bytes_count(&self) -> usize {
+        self.state.sync_state.buffer.len()
+    }
+
+    /// Process a new byte during a synchronized update.
+    #[cold]
+    fn advance_sync<H, W>(&mut self, handler: &mut H, byte: u8, writer: &mut W)
+    where
+        H: Handler,
+        W: io::Write,
+    {
+        self.state.sync_state.buffer.push(byte);
+
+        // Handle sync DCS escape sequences.
+        match self.state.sync_state.pending_dcs {
+            Some(_) => self.advance_sync_dcs_end(handler, byte, writer),
+            None => self.advance_sync_dcs_start(),
+        }
+    }
+
+    /// Find the start of sync DCS sequences.
+    fn advance_sync_dcs_start(&mut self) {
+        // Get the last few bytes for comparison.
+        let len = self.state.sync_state.buffer.len();
+        let offset = len.saturating_sub(SYNC_ESCAPE_START_LEN);
+        let end = &self.state.sync_state.buffer[offset..];
+
+        // Check for extension/termination of the synchronized update.
+        if end == SYNC_START_ESCAPE_START {
+            self.state.sync_state.pending_dcs = Some(Dcs::SyncStart);
+        } else if end == SYNC_END_ESCAPE_START || len >= SYNC_BUFFER_SIZE - 1 {
```

Comment:
I think there's still an issue with intermediate parameters in the end sequence. For example, I think I'd expect this to sync, then stop, however instead I hit the timeout before my next prompt:

```sh
echo -en "\eP=1s\e\x5c"; echo -en "\eP=2;foo\e\x5c";
```

Reply:
That's a good question actually and I'm not entirely sure on that one. The spec mentions additional parameters should be ignored, but I'm not certain they're talking about the passthrough or the dcs parameters.

It should likely be allowed to pass additional dcs params, in which case your escape sequence should be valid and the second `0` parameter should just be ignored.

Reply:
`0` parameter?

Reply:
If you add a semicolon, delimiting parameters, without specifying any exact value, it defaults to `0`. So in your case it's basically a list of parameters 2 and 0, then the `foo` passthrough content.

Reply:
I see, I haphazardly assumed I could pass `foo` as a parameter, then noticed the behavior.

Reply:
According to the specification for this feature, all DCS parameters need to be accepted and ignored. However the spec refers to the DCS passthrough as parameters:

https://gitlab.freedesktop.org/terminal-wg/specifications/-/merge_requests/2/diffs#8479248bbb91ebb045ead1bea64cc581ee240bba_0_65

Since there is no mention of the actual DCS params being ignored (the state *between* `1` and `s`), I'd assume that no other parameter is valid in this position and thus should cause an invalidation, which means ignoring the escape is correct (or at least acceptable).

Reply:
I see, I mixed up the pass-throughs and the parameters. This seems to be working as intended.